File: src/main/kotlin/java/infrastructure/wisesaying/WiseSayingRepositoryImpl.java
/*
package java.infrastructure.wisesaying;

import com.infrastructure.wisesaying.entity.WiseSayingEntity;
import com.infrastructure.wisesaying.repository.WiseSayingRepository;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

import wisesaying.domain.WiseSaying;
import wisesaying.exception.WiseSayingException;

public class WiseSayingRepositoryImpl implements WiseSayingRepository {

	private final IdGeneration idGeneration;
	private final LinkedHashMap<Long, WiseSayingEntity> wiseSayingEntityLinkedHashMap;
	private static final String FILE_PATH = "db/wiseSaying/";
	private static final String LAST_ID_PATH = FILE_PATH + "lastId.txt";

	public WiseSayingRepositoryImpl() {
		Path lastIdPath = Paths.get(LAST_ID_PATH);
		Path wiseSayingPath = Paths.get(FILE_PATH + "data.json");

		try {
			if (Files.exists(lastIdPath)) {
				long id = Long.parseLong(Files.readString(lastIdPath));
				idGeneration = new IdGeneration(id);
			} else {
				idGeneration = new IdGeneration(0L);
			}

			if (Files.exists(wiseSayingPath)) {
				String json = Files.readString(wiseSayingPath);
				wiseSayingEntityLinkedHashMap = WiseSayingEntity.fromJsonList(json);
			} else {
				wiseSayingEntityLinkedHashMap = new LinkedHashMap<>();
			}
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	public Long add(WiseSaying wiseSaying) throws IOException {
		Long id = idGeneration.generationId();
		WiseSayingEntity wiseSayingEntity = new WiseSayingEntity(wiseSaying.getWiseSaying(),
			wiseSaying.getWriter());
		wiseSayingEntity.setId(id);
		wiseSayingEntityLinkedHashMap.put(id, wiseSayingEntity);

		Path wiseSayingPath = Paths.get(FILE_PATH + wiseSayingEntity.getId() + ".json");
		Path lastIdPath = Paths.get(LAST_ID_PATH);

		if (!Files.exists(wiseSayingPath)) {
			Files.createDirectories(wiseSayingPath.getParent());
			Files.createFile(wiseSayingPath);
		}

		if (!Files.exists(lastIdPath)) {
			Files.createFile(lastIdPath);
		}

		String wiseSayingEntityJson = wiseSayingEntity.toJson();

		Files.write(wiseSayingPath, wiseSayingEntityJson.getBytes());
		Files.write(lastIdPath, String.valueOf(id).getBytes());

		return id;
	}

	public Optional<WiseSaying> findById(Long id) throws IOException {
		WiseSayingEntity wiseSayingEntity = wiseSayingEntityLinkedHashMap.get(id);

		if (Objects.isNull(wiseSayingEntity)) {
			return Optional.empty();
		}

		WiseSaying wiseSaying = new WiseSaying(
			wiseSayingEntity.getId(),
			wiseSayingEntity.getWiseSaying(),
			wiseSayingEntity.getWriter()
		);

		return Optional.of(wiseSaying);
	}

	public Optional<LinkedList<WiseSaying>> findAll() throws IOException {
		if (wiseSayingEntityLinkedHashMap.isEmpty()) {
			return Optional.of(new LinkedList<>());
		}

		return Optional.of(wiseSayingEntityLinkedHashMap.values().stream()
			.map(WiseSayingEntity::toModel)
			.collect(Collectors.toCollection(LinkedList::new)));
	}

	public Long delete(Long id) throws IOException {
		Path wiseSayingPash = Paths.get(FILE_PATH + id + ".json");
		WiseSayingEntity remove = wiseSayingEntityLinkedHashMap.remove(id);

		if (Objects.isNull(remove) || !Files.exists(wiseSayingPash)) {
			throw new WiseSayingException(id + "번 명언은 존재하지 않습니다.");
		}

		Files.delete(wiseSayingPash);
		return id;
	}

	public void update(WiseSaying wiseSaying) throws IOException {
		WiseSayingEntity wiseSayingEntity = WiseSayingEntity.from(wiseSaying);
		Path wiseSayingPath = Paths.get(FILE_PATH + wiseSayingEntity.getId() + ".json");
		wiseSayingEntityLinkedHashMap.replace(wiseSayingEntity.getId(), wiseSayingEntity);
		Files.write(wiseSayingPath, wiseSayingEntity.toJson().getBytes());
	}

	public Boolean build() throws IOException {
		Path wiseSayingPath = Paths.get(FILE_PATH + "data.json");
		if (wiseSayingEntityLinkedHashMap.isEmpty()) {
			return Boolean.FALSE;
		}

		if (!Files.exists(wiseSayingPath)) {
			Files.createDirectories(wiseSayingPath.getParent());
			Files.createFile(wiseSayingPath);
		}

		Files.write(wiseSayingPath,
			WiseSayingEntity.toJsonList(wiseSayingEntityLinkedHashMap).getBytes());

		return Boolean.TRUE;
	}
}
*/

File: src/main/kotlin/infrastructure/wisesaying/repository/WiseSayingRepositoryImpl.kt
package com.infrastructure.wisesaying.repository

import com.infrastructure.wisesaying.entity.IdGeneration
import com.infrastructure.wisesaying.entity.WiseSayingEntity
import com.wisesaying.mapper.WiseSayingMapper
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*
import java.util.stream.Collectors
import java.wisesaying.domain.WiseSaying
import kotlin.collections.LinkedHashMap
import kotlin.io.path.Path

class WiseSayingRepositoryImpl : WiseSayingRepository {
    private val idGeneration: IdGeneration
    private val wiseSayingEntityLinkedHashMap: LinkedHashMap<Long, WiseSayingEntity>

    companion object {
        private const val FILE_PATH = "db/wiseSaying/"
        private const val LAST_ID_PATH = "${FILE_PATH}lastId.txt"
    }

    init {
        val lastIdPath = Paths.get(LAST_ID_PATH)
        val wiseSayingPath = Paths.get("${FILE_PATH}data.json")

        try {
            idGeneration = if (Files.exists(lastIdPath)) {
                val id = Files.readString(lastIdPath).toLong()
                IdGeneration(id)
            } else {
                IdGeneration(0L)
            }

            wiseSayingEntityLinkedHashMap = if (Files.exists(wiseSayingPath)) {
                val json = Files.readString(wiseSayingPath)
                WiseSayingMapper.toLinkedHashMap(json)

            } else {
                LinkedHashMap()
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    override fun save(wiseSaying: WiseSaying): Long {
        val id = idGeneration.generationId();
        val saveWiseSayingEntity = WiseSayingEntity(id, wiseSaying.wiseSaying, wiseSaying.writer)

        val wiseSayingPath = getWiseSayingPath(saveWiseSayingEntity.id)

        val lastIdPath = Paths.get(LAST_ID_PATH);

        if (!Files.exists(wiseSayingPath)) {
            Files.createDirectories(wiseSayingPath.parent)
            Files.createFile(wiseSayingPath)
        }

        if (!Files.exists(lastIdPath)) Files.createFile(wiseSayingPath)

        val saveWiseSayingEntityJson = WiseSayingMapper.toJson(saveWiseSayingEntity)

        try {
            Files.write(wiseSayingPath, saveWiseSayingEntityJson.toByteArray())
            Files.write(lastIdPath, saveWiseSayingEntity.id.toString().toByteArray())

            wiseSayingEntityLinkedHashMap[saveWiseSayingEntity.id] = saveWiseSayingEntity
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        return saveWiseSayingEntity.id;
    }

    override fun findById(id: Long): Optional<WiseSaying> {
        val findWiseSaying = wiseSayingEntityLinkedHashMap[id] ?: return Optional.empty();

        return Optional.of(
            WiseSaying(findWiseSaying.id, findWiseSaying.wiseSaying, findWiseSaying.writer)
        )
    }

    override fun delete(id: Long): Long {
        val wiseSayingPath = getWiseSayingPath(id)
        val remove = wiseSayingEntityLinkedHashMap.remove(id) ?: throw RuntimeException("예외")

        Files.delete(wiseSayingPath)

        return remove.id
    }

    override fun update(wiseSaying: WiseSaying) {
        val updateWiseSayingEntity =
            WiseSayingEntity(wiseSaying.id, wiseSaying.wiseSaying, wiseSaying.writer)

        val wiseSayingPath = getWiseSayingPath(updateWiseSayingEntity.id)

        try {
            Files.write(wiseSayingPath, updateWiseSayingEntity.toJson().toByteArray())
            wiseSayingEntityLinkedHashMap[updateWiseSayingEntity.id] = updateWiseSayingEntity
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    override fun findAll(): LinkedList<WiseSaying> {
        if (wiseSayingEntityLinkedHashMap.isEmpty()) return LinkedList();

        return wiseSayingEntityLinkedHashMap.values
            .map(WiseSayingMapper::toDomain).toCollection(LinkedList())
    }

    override fun build(): Boolean {
        val buildPath = Paths.get("${FILE_PATH}data.jaon")

        if (wiseSayingEntityLinkedHashMap.isEmpty()) return false

        if (!Files.exists(buildPath)) {
            Files.createDirectories(buildPath.parent)
            Files.createFile(buildPath)
        }

        Files.write(buildPath, WiseSayingMapper.toJson(wiseSayingEntityLinkedHashMap).toByteArray())

        return true
    }

    private fun getWiseSayingPath(id: Long): Path = Paths.get("${FILE_PATH}${id}.json")
}